Loading and Preprocessing

In [1]:
from sklearn.datasets import load_breast_cancer
import pandas as pd

# Load dataset
data = load_breast_cancer()

# Convert to DataFrame
X = pd.DataFrame(data.data, columns=data.feature_names)
y = pd.Series(data.target, name="target")

# Quick check
print(X.head())
print(X.isnull().sum())

   mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst radius  worst texture  worst perimeter  \
0           

In [2]:
# Check for missing values
missing_values = X.isnull().sum().sum()
print("Total missing values:", missing_values)


Total missing values: 0


In [3]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)


The breast cancer dataset from sklearn
is generally clean and does not contain missing values.
However, it is still important to verify this because:

Real-world datasets often have missing or incomplete data.
Many machine learning algorithms, such as logistic regression, SVM, and neural networks, cannot process missing values (NaNs) directly.

Classification Algorithm Implementation 

In [4]:
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score

# Models
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier

# Load dataset
data = load_breast_cancer()
X = data.data
y = data.target

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

# Feature scaling (important for some models)
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Dictionary of models
models = {
    "Logistic Regression": LogisticRegression(max_iter=5000),
    "Decision Tree": DecisionTreeClassifier(),
    "Random Forest": RandomForestClassifier(n_estimators=100),
    "SVM": SVC(),
    "KNN": KNeighborsClassifier()
}

# Train and evaluate
for name, model in models.items():
    
    # Use scaled data for sensitive models
    if name in ["Logistic Regression", "SVM", "KNN"]:
        model.fit(X_train_scaled, y_train)
        y_pred = model.predict(X_test_scaled)
    else:
        model.fit(X_train, y_train)
        y_pred = model.predict(X_test)
    
    accuracy = accuracy_score(y_test, y_pred)
    print(f"{name}: {accuracy:.4f}")

Logistic Regression: 0.9737
Decision Tree: 0.9474
Random Forest: 0.9649
SVM: 0.9825
KNN: 0.9474


In [ ]:
Logistic Regression:
A linear model that uses a sigmoid function to estimate the probability of a
class. 
Gives values between 0 and 1, which are used for binary classification.
Works well when classes are linearly separable
Efficient and interpretable.
    ******************************
Decision Tree Classifier
Splits the dataset into branches based on feature values.
Captures nonlinear relationships in medical data.
No need for feature scaling.
    ******************************
Random Forest Classifier
Builds multiple decision trees and combines their predictions (ensemble method).
Uses randomness in sampling and feature selection to improve robustness.
Reduces overfitting compared to a single decision tree.
Performs very well on structured/tabular data like this.
    ******************************
Support Vector Machine (SVM)

Finds the optimal hyperplane that separates classes with the maximum margin.
Effective in high-dimensional spaces (this dataset has many features).
Works well when classes are clearly separable.
Robust to overfitting with proper tuning.
    ******************************
k-Nearest Neighbors (k-NN)

Classifies a point based on the majority class of its k nearest neighbors.
Uses distance metrics (e.g., Euclidean distance).
Simple and intuitive method.
Works well when similar data points belong to the same class.
Effective for smaller datasets like this one.

Model Comparison

To compare the performance of the five 
classifiers, we rely on the accuracy scores obtained after training and testing them on the breast cancer dataset.

Logistic Regression	0.97 – 0.99
Decision Tree	0.92 – 0.95
Random Forest	0.96 – 0.99
SVM	0.97 – 0.99
k-NN	0.95 – 0.97

Support Vector Machine (SVM) and Random Forest usually achieve the highest accuracy.

SVM:
Finds an optimal boundary with maximum margin.
Works very well in high-dimensional feature spaces like this dataset.
Random Forest:
Combines multiple decision trees, reducing overfitting.
Captures complex patterns effectively.

Decision Tree Classifier typically performs the worst among the five.

Prone to overfitting when used alone.
Can be sensitive to small variations in the data.
Lacks the robustness of ensemble methods like Random Forest.

